File: leetcode/97/src/P97.java
/**
 * Created by qwordy on 2/22/16.
 * 97. Interleaving String
 */

class Solution {
	public boolean isInterleave(String s1, String s2, String s3) {
		int n, m, i, j;
		n = s1.length();
		m = s2.length();
		if (n + m != s3.length())
			return false;
		boolean[][] d = new boolean[n + 1][m + 1];
		d[0][0] = true;
		for (i = 1; i <= n && d[i - 1][0]; i++)
			d[i][0] = s1.charAt(i - 1) == s3.charAt(i - 1);
		for (j = 1; j <= m && d[0][j - 1]; j++)
			d[0][j] = s2.charAt(j - 1) == s3.charAt(j - 1);
		for (i = 1; i <= n; i++)
			for (j = 1; j <= m; j++)
				d[i][j] = d[i - 1][j] && s1.charAt(i - 1) == s3.charAt(i + j - 1)
						|| d[i][j - 1] && s2.charAt(j - 1) == s3.charAt(i + j - 1);
		return d[n][m];
	}
}

public class P97 {
}

File: leetcode/97/P97.rb
# 97. Interleaving String
# @param {String} s1
# @param {String} s2
# @param {String} s3
# @return {Boolean}
def is_interleave(s1, s2, s3)
	n = s1.size
	m = s2.size
	if n + m != s3.size
		return false
	end
	d = Array.new(n + 1) { Array.new m + 1, false }
	d[0][0] = true
	for i in 1..n
		break if !(d[i][0] = s1[i - 1] == s3[i - 1])
	end
	for j in 1..m
		break if !(d[0][j] = s2[j - 1] == s3[j - 1])
	end
	for i in 1..n
		for j in 1..m
			d[i][j] = d[i - 1][j] && s1[i - 1] == s3[i + j - 1] \
					|| d[i][j - 1] && s2[j - 1] == s3[i + j - 1]
		end
	end
	return d[n][m]
end

p is_interleave 'db', 'b', 'cbb'

